runtime/src/common/sgx/avr.rs

```diff
@@ -342,47 +326,156 @@ fn validate_avr_signature(
+    //
+    //  * 4.2.2 Report Signature:
+    //    * "The Attestation Verification Report is cryptographically
+    //       signed by Report Signing Key (owned by the Attestation
+    //       Service) using the RSA-SHA256 algorithm."
+    //  * 4.2.3 Report Signing Certificate Chain:
+    //    * "The public part of Report Key is distributed in the form
+    //       of an x.509 digital certificate called Attestation Report
+    //       Signing Certificate. It is a leaf certificate issued by
+    //       the Attestation Report Signing CA Certificate"
+    //    * "A PEM-encoded certificate chain consisting of Attestation
+    //       Report Signing Certificate and Attestation Report Signing
+    //       CA Certificate is returned..."
+    //
+    // See: "Attestation Service for Intel(R) Software Guard Extensions
+    // (Intel(R) SGX): API Documentation" (Revision: 6.0)
 
-    // Decode the certificate chain.
+    // Decode the certificate chain from percent encoded PEM to DER.
     let raw_pem = percent_encoding::percent_decode(cert_chain).decode_utf8()?;
-    let mut cert_chain = Vec::new();
-    for pem in parse_many(&raw_pem.as_bytes()) {
-        if pem.tag != PEM_CERTIFICATE_LABEL {
+    let mut cert_ders = Vec::new();
+    for pem in pem::Pem::iter_from_buffer(&raw_pem.as_bytes()) {
+        let pem = match pem {
+            Ok(p) => p,
+            Err(_) => return Err(AVRError::MalformedCertificatePEM.into()),
+        };
+        if pem.label != PEM_CERTIFICATE_LABEL {
             return Err(AVRError::MalformedCertificatePEM.into());
         }
-        cert_chain.push(pem.contents);
+        cert_ders.push(pem.contents);
     }
-    if cert_chain.len() == 0 {
-        return Err(AVRError::NoCertificates.into());
+
+    // IAS per the API will only ever send two certificates.
+    if cert_ders.len() != 2 {
+        return Err(AVRError::ChainNotTwoCertificates.into());
     }
 
-    // Decode the signature.
-    let signature = base64::decode(signature)?;
+    // Convert our timestamp to something that can be used to check
+    // certificate expiration.
+    let time = ASN1Time::from_timestamp(unix_time as i64);
 
-    let time = webpki::Time::from_seconds_since_unix_epoch(unix_time);
+    // Attestation Report Signing CA Certificate:
+    //
+    // Ensure that it matches the hard-coded copy, and decode it, so
+    // that the expiration can be validated and the public key can
+    // be used to verify the leaf certificate's signature.
+    //
+    // This could be more paranoid and check that the cert doesn't
+    // have trailing garbage, the usage is correct, etc, but we can
+    // take it as a matter of faith that it is well-formed since
+    // it is the same as the hard-coded one.
+    //
+    // TODO/perf: In theory this can be done once and only once, but
+    // the borrow checker thwarted my attempts to initialize a tuple
+    // containing a X509Certificate and Pem via lazy_static.
```

Comment:
What was the issue here?

Reply:
The X509Certificate doesn't copy/move out of the Pem, so the borrow checker got mad.